src/polynomial/grouped_poly.rs

```diff
@@ -27,27 +26,55 @@ impl<F> Default for GroupedPoly<F> {
     }
 }
 
+impl<F: PrimeField> From<HashMap<usize, Expression<F>>> for GroupedPoly<F> {
+    fn from(value: HashMap<usize, Expression<F>>) -> Self {
```

Comment:
Do you still plan to use HashMap but only convert HashMap to vector at the end?

Reply:
I changed to the most generalised implementation that allows you to create it from any map or iterator, now I think there should be no question about it

Now it's `impl IntoIter<Item = (usize, Expression<F>)>`

Reply:
Since we use Vec<Expression> to represent the GroupedPoly, it seems that there is no need to do this conversion:
```
fn from(value: IT) -> Self {
        let mut self_ = Self::default();

        for (degree, expr) in value.into_iter() {
            if degree >= self_.terms.len() {
                self_.terms.resize(degree + 1, None);
            }

            self_.terms[degree] = Some(expr);
        }

        self_
    }
```
since the index of Vec<Expression>  is the degree. Do you have any special use case to use the type `Iterator<(degree, Expression)>` instead of `Vec<Expression>` that I am not aware of?

Reply:
Well, all our tests represent this case, when we don't have an expression for e.g. degree 2, but we have one for degree 3.

I.e. it's alternative to create "as is" via `Vec<Option<Expression>>`, then 
```rust
GroupedPoly::<Fq>::from(map! {
        2 => Expression::Polynomial(Query { index: 0, rotation: Rotation(0) }),
        3 => Expression::Polynomial(Query { index: 1, rotation: Rotation(0) }),
        4 => Expression::Polynomial(Query { index: 2, rotation: Rotation(0) }),
}
```

```rust
GroupedPoly::<Fq>::from(vec![
    None, // 0
    None, // 1
    Some(Expression::Polynomial(Query { index: 0, rotation: Rotation(0) })), // 2
    Some(Expression::Polynomial(Query { index: 1, rotation: Rotation(0) })), // 3
    Some(Expression::Polynomial(Query { index: 2, rotation: Rotation(0) })), // 4
}
```

The second option reads an order of magnitude worse in my opinion.

Reply:
The alternative option is to use `Expression::Const(F::ZERO)`, but it seems to me that this would still read worse, especially if we were to describe test cases with a large degree somewhere else

Reply:
With `IntoIter` this code will work too
```rust
GroupedPoly::<Fq>::from([
    (0, Expression::Constant(Fq::from_u128(u128::MAX))),
    (
        1,
        Expression::Polynomial(Query {
            index: 0,
            rotation: Rotation(0),
        }),
    ),
    (5, Expression::Challenge(0)),
])
```

without any macro from maplit, it's just a little less readable because of the formatting, but not critical

Reply:
IC, it will simplify the creation of GroupedPoly.